build-scripts/build_tests.py

```diff
@@ -0,0 +1,218 @@
+#!/usr/bin/python3
+
+import argparse
+import logging
+import pathlib
+import sys
+from typing import TypeVar, Iterator, List, Iterable, Set, Dict
+import multiprocessing
+
+import ssg.environment
+import ssg.jinja
+import ssg.utils
+import ssg.yaml
+import ssg.templates
+import tests.ssg_test_suite.common
+import tests.ssg_test_suite.rule
+
+SSG_ROOT = str(pathlib.Path(__file__).resolve().parent.parent.absolute())
+JOB_COUNT = multiprocessing.cpu_count()
+T = TypeVar('T')
+
+
+def _create_arg_parser() -> argparse.ArgumentParser:
+    parser = argparse.ArgumentParser(
+        description="Converts builds content tests to be rendered"
+    )
+    parser.add_argument(
+        "--build-config-yaml", required=True,
+        help="YAML file with information about the build configuration. "
+             "e.g.: ~/scap-security-guide/build/build_config.yml"
+    )
+    parser.add_argument(
+        "--product-yaml", required=True,
+        help="YAML file with information about the product we are building. "
+             "e.g.: ~/scap-security-guide/rhel10/product.yml"
+    )
+    parser.add_argument(
+        "--output", required=True,
+        help="Output path"
+             "e.g.:  ~/scap-security-guide/build/rhel10/tests"
+    )
+    parser.add_argument(
+        "--resolved-rules-dir", required=True,
+        help="Directory with <rule-id>.yml resolved rule YAMLs"
+             "e.g.: ~/scap-security-guide/build/rhel10/rules"
+    )
+    parser.add_argument("--verbose", action="store_true", default=False)
+    parser.add_argument("--root", default=SSG_ROOT,
+                        help=f"Path to the project. Defaults to {SSG_ROOT}")
+    parser.add_argument("--jobs", "-j", type=int, default=JOB_COUNT,
+                        help=f"Number of cores to use. Defaults to {JOB_COUNT} on this system.")
+    return parser
+
+
+def _is_test_file(filename: str) -> bool:
+    return (filename.endswith('.pass.sh')
+            or filename.endswith('.fail.sh') or
+            filename.endswith('.notapplicable.sh'))
+
+
+def _divide_iters(iterator: Iterable[T], n: int) -> List[Iterator[T]]:
+    items = list(iterator)
+    length = len(items)
+
+    chunk_size = length // n
+    remainder = length % n
+
+    inters = []
+    start = 0
+
+    for i in range(n):
+        end = start + chunk_size + (1 if i < remainder else 0)
+        inters.append(iter(items[start:end]))
+        start = end
+    return inters
```

Comment:
If you don't mind interleaving the split, you could use python's array slicing "step" property.

Given 3 workers and a total of 10 rules, you can assign rules to them like this:
```python
>>> a = [1,2,3,4,5,6,7,8,9,10]
>>> a[0::3]
[1, 4, 7, 10]
>>> a[1::3]
[2, 5, 8]
>>> a[2::3]
[3, 6, 9]
```
where the first (`start`) number is worker index, and the second (`step`) number is the total number of workers.

This is what we use in Contest to avoid any off-by-one problems and ensure no rule is left behind, while making things just simpler and evenly distributing rule complexity across all workers (as complex/simple rules tend to alphabetically bunch up).

https://github.com/RHSecurityCompliance/contest/blob/1204f812956a855eb15d26b1014cb28e5306ee05/per-rule/test.py#L93-L97